blob/azureblob/azureblob.go

```diff
@@ -0,0 +1,586 @@
+package azureblob
+
+import (
+	"bytes"
+	"context"	
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+
+	"github.com/google/go-cloud/blob"
+	"github.com/google/go-cloud/blob/driver"
+
+	"github.com/Azure/azure-pipeline-go/pipeline"
+	"github.com/Azure/azure-storage-blob-go/azblob"
+)
+
+type bucket struct {
+	name             string
+	credential       *azblob.SharedKeyCredential // used for SignedURL
+	PublicAccessType azblob.PublicAccessType
+	urls             serviceUrls
+	pageMarkers      map[string]azblob.Marker // temporary page marker map, until azblob.Marker is an exportable type
+	defaultDelimiter string                   // for escaping backslashes
+}
+
+type serviceUrls struct {
+	serviceURL   *azblob.ServiceURL   // represents the Azure Storage Account
+	containerURL *azblob.ContainerURL // represents the Azure Storage Container
+	blockBlobURL *azblob.BlockBlobURL // represents the Azure Block Blob
+}
+
+// Settings to establish connection to Azure
+type Settings struct {
+	AccountName      string
+	AccountKey       string
+	PublicAccessType azblob.PublicAccessType
+	SASToken         string
+	Pipeline         pipeline.Pipeline
+	DefaultDelimiter DefaultBlobDelimiter
+}
+
+// DefaultBlobDelimiter is used to escape backslashes
+type DefaultBlobDelimiter struct {
+	Value string
+}
+
+// See https://msdn.microsoft.com/en-us/library/azure/dd179468.aspx and "x-ms-blob-public-access" header.
+const (
+	// PublicAccessBlob
+	PublicAccessBlob azblob.PublicAccessType = azblob.PublicAccessBlob
+	// PublicAccessContainer
+	PublicAccessContainer azblob.PublicAccessType = azblob.PublicAccessContainer
+	// PublicAccessNone represents an empty PublicAccessType.
+	PublicAccessNone azblob.PublicAccessType = azblob.PublicAccessNone
+)
+
+const (
+	// BlobPathSeparator is used to escape backslashes
+	BlobPathSeparator = "/"
+	// OSPathSeparator or backslashes must be converted to forwardslashes
+	OSPathSeparator = string(os.PathSeparator)
+)
+
+var (
+	maxDownloadRetryRequests = 3    // download retry policy
+	maxPageSize              = 5000 // default page size for ListPaged
+)
+
+// OpenBucket returns an Azure BlockBlob Bucket
+func OpenBucket(ctx context.Context, settings *Settings, containerName string) (*blob.Bucket, error) {
```

Comment:
I hear what you are saying, and it's a good goal.

However, I don't think it's possible to hide all of the Azure stuff during initialization. There are just too many configuration options, etc. to make it feasible -- you can see how much it has complicated your code, and I expect you're only exposing a small subset of the ways it can actually be done. By trying to create them inside the API, you are preventing users from doing it themselves. As described below, it works better to give them the flexibility, but then provide helpers/tools that make the common case easy.

To help with the initialization phase, we have a dependency injection tool called `Wire`. For gcp and aws we have top-level packages in the repo with some easy-to-use helpers if you just want the defaults, and Wire `ProviderSets` with them to make it easy to inject them. Some of the code you have here should probably go into an `azure` directory that does the same (in a separate PR please).

Go Cloud's value add is not during the provisioning or initialization phase, it's that once you've done that you can write your application code once.

I hope that's convincing....